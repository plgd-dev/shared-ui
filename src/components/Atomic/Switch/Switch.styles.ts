import { css } from '@emotion/react'
import { SwitcherSizeType } from './Switch.types'
import { colors } from '../_utils/colors'
import styled from '@emotion/styled'
import { ThemeType, getThemeColor, getTheme } from '../_theme'
import { fontPrimary } from '../_utils/commonStyles'

export const switchC = css`
    display: inline-flex;
    align-items: center;
`

export const labelBefore = css`
    flex-direction: row-reverse;
`

const getSwitcherSize = (size: SwitcherSizeType) => css`
    width: ${size === 'big' ? '32px' : '24px'};
    height: ${size === 'big' ? '20px' : '16px'};
    border-radius: ${size === 'big' ? '26.6667px' : '20px'};
    overflow: hidden;
`

export const switcher = (size: SwitcherSizeType) => css`
    ${getSwitcherSize(size)};
    background: ${colors.neutral300};
    position: relative;
`

export const slider = styled.span`
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: ${(props) => getThemeColor(props.theme, `Switch.background`)};
    transition: 0.4s;

    &:before {
        position: absolute;
        content: '';
        left: 2px;
        bottom: 2px;
        background-color: white;
        transition: 0.4s;
        border-radius: 50%;
        width: ${(props: any) => (props.size === 'big' ? '16px' : '12px')};
        height: ${(props: any) => (props.size === 'big' ? '16px' : '12px')};
    }
`

export const input = (theme: ThemeType, size: SwitcherSizeType, disabled: boolean) => css`
    position: absolute;
    opacity: 0;
    width: 0;
    height: 0;

    &:checked ~ ${slider} {
        background: ${disabled ? getThemeColor(theme, 'Switch.input.background.disabled') : getThemeColor(theme, 'Switch.input.background.primary')};

        &:before {
            transform: translateX(${size === 'big' ? '12px' : '8px'});
        }
    }
`

export const loading = css`
    & ~ ${slider} {
        background: ${colors.neutral500}!important;
    }
`

export const sliderStyle = (size: SwitcherSizeType) => css`
    ${getSwitcherSize(size)};

    &:before {
        width: ${size === 'big' ? '16px' : '12px'};
        height: ${size === 'big' ? '16px' : '12px'};
    }
`

export const label = (theme: ThemeType) => css`
    font-size: 14px;
    color: ${getThemeColor(theme, `Switch.label.color`)};
    padding-left: 8px;
    cursor: pointer;
    line-height: 18px;
    font-family: ${getTheme(theme, `Global.fontPrimary`, fontPrimary)};
    font-style: normal;
    font-weight: 400;
`

export const labelBeforeSwitch = css`
    padding-left: 0;
    padding-right: 8px;
`

export const disabled = css`
    cursor: not-allowed !important;
`
