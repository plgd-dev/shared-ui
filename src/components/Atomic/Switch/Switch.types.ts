export type SwitcherSizeType = 'small' | 'big'

export type Props = {
    checked?: boolean
    className?: string
    defaultChecked?: boolean
    disabled?: boolean
    id?: string
    label?: string
    labelBefore?: boolean
    loading?: boolean
    name?: string
    onChange?: (e: any) => void
    size?: SwitcherSizeType
}

export const defaultProps: Partial<Props> = {
    labelBefore: false,
    size: 'big',
}
