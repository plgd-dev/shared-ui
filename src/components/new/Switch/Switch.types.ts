export type Props = {
    checked?: boolean
    disabled?: boolean
    id: string
    label: string
    onChange: (e: any) => void
}
