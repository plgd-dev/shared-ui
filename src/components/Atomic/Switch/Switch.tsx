import React, { forwardRef } from 'react'
import { Props, defaultProps } from './Switch.types'
import * as styles from './Switch.styles'

const Switch = forwardRef<HTMLInputElement, Props>((props, ref) => {
    const { id, label, labelBefore, size, defaultChecked, className, disabled, loading, ...rest } = { ...defaultProps, ...props }
    const Slider = styles.slider as any

    return (
        <label className={className} css={[styles.switchC, labelBefore && styles.labelBefore]}>
            <div css={styles.switcher(size || 'big')}>
                <input
                    {...rest}
                    css={(theme) => [styles.input(theme, size || 'big', disabled || false), loading && styles.loading]}
                    defaultChecked={defaultChecked}
                    disabled={disabled || loading}
                    ref={ref}
                    type='checkbox'
                />
                <Slider css={[styles.sliderStyle(size || 'big'), disabled && styles.disabled]} size={size} />
            </div>
            {label && <div css={[styles.label, labelBefore && styles.labelBeforeSwitch]}>{label}</div>}
        </label>
    )
})

Switch.displayName = 'Switch'

export default Switch
