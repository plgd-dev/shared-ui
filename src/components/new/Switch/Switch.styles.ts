import { css } from '@emotion/react'
import { SwitcherSizeType } from '@plgd/shared-ui/components/new/Switch/Switch.types'
import { colors } from '../_utils/colors'
import styled from '@emotion/styled'

export const switchC = css`
    display: flex;
    align-items: center;
`

const getSwitcherSize = (size: SwitcherSizeType) =>
    css`
        width: ${size === 'big' ? '32px' : '24px'};
        height: ${size === 'big' ? '20px' : '15px'};
        border-radius: ${size === 'big' ? '26.6667px' : '20px'};
    `

export const switcher = (size: SwitcherSizeType) => css`
    ${getSwitcherSize(size)};
    background: ${colors.neutral300};
    position: relative;
`

export const slider = styled.span`
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: #ccc;
    transition: 0.4s;

    &:before {
        position: absolute;
        content: '';
        left: 2px;
        bottom: 2px;
        background-color: white;
        transition: 0.4s;
        border-radius: 50%;
    }
`

export const input = (size: SwitcherSizeType) => css`
    opacity: 0;
    width: 0;
    height: 0;

    &:checked ~ ${slider} {
        background: ${colors.primary};

        &:before {
            transform: translateX(${size === 'big' ? '12px' : '8px'});
        }
    }
`

export const sliderStyle = (size: SwitcherSizeType) => css`
    ${getSwitcherSize(size)};

    &:before {
        width: ${size === 'big' ? '16px' : '11px'};
        height: ${size === 'big' ? '16px' : '11px'};
    }
`

export const label = css`
    font-size: 12px;
    color: #81868c;
    padding-left: 8px;
    cursor: pointer;
    line-height: 18px;
    font-family: 'Poppins', sans-serif;
    font-style: normal;
    font-weight: 400;
`
