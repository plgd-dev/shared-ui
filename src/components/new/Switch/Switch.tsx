import React, { FC } from 'react'
import { Props, defaultProps } from './Switch.types'
import * as styles from './Switch.styles'

const Switch: FC<Props> = (props) => {
    const { id, label, labelBefore, size, defaultChecked, className, disabled, ...rest } = { ...defaultProps, ...props }
    const Slider = styles.slider

    return (
        <label css={[styles.switchC, labelBefore && styles.labelBefore]} className={className}>
            <div css={styles.switcher(size || 'big')}>
                <input {...rest} css={styles.input(size || 'big', disabled || false)} defaultChecked={defaultChecked} type='checkbox' disabled={disabled} />
                <Slider css={[styles.sliderStyle(size || 'big'), disabled && styles.disabled]} />
            </div>
            {label && <div css={[styles.label, labelBefore && styles.labelBeforeSwitch]}>{label}</div>}
        </label>
    )
}

Switch.displayName = 'Switch'

export default Switch
