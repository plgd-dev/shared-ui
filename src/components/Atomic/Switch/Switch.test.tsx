import React from 'react'
import { render } from '../../../../test/jest-wrapper'
import Switch from './Switch'

describe('<Switch>', () => {
    it('render correctly - snapshot', () => {
        const { asFragment } = render(
            <div>
                <Switch size='small' />
                <Switch size='big' />
                <Switch defaultChecked={true} size='small' />
                <Switch defaultChecked={true} size='big' />
                <Switch defaultChecked={true} label='Label text' size='big' />
            </div>
        )

        expect(asFragment()).toMatchSnapshot()
    })
})
