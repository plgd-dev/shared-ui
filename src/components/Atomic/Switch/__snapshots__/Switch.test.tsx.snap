// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  width: 24px;
  height: 16px;
  border-radius: 20px;
  overflow: hidden;
  background: #D7D8DA;
  position: relative;
}

.emotion-2 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
}

.emotion-2:checked~.emotion-4 {
  background: #2261AE;
}

.emotion-2:checked~.emotion-4:before {
  -webkit-transform: translateX(8px);
  -moz-transform: translateX(8px);
  -ms-transform: translateX(8px);
  transform: translateX(8px);
}

.emotion-3 {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #D7D8DA;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  width: 24px;
  height: 16px;
  border-radius: 20px;
  overflow: hidden;
}

.emotion-3:before {
  position: absolute;
  content: '';
  left: 2px;
  bottom: 2px;
  background-color: white;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  border-radius: 50%;
  width: 12px;
  height: 12px;
}

.emotion-3:before {
  width: 12px;
  height: 12px;
}

.emotion-6 {
  width: 32px;
  height: 20px;
  border-radius: 26.6667px;
  overflow: hidden;
  background: #D7D8DA;
  position: relative;
}

.emotion-7 {
  position: absolute;
  opacity: 0;
  width: 0;
  height: 0;
}

.emotion-7:checked~.emotion-4 {
  background: #2261AE;
}

.emotion-7:checked~.emotion-4:before {
  -webkit-transform: translateX(12px);
  -moz-transform: translateX(12px);
  -ms-transform: translateX(12px);
  transform: translateX(12px);
}

.emotion-8 {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: #D7D8DA;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  width: 32px;
  height: 20px;
  border-radius: 26.6667px;
  overflow: hidden;
}

.emotion-8:before {
  position: absolute;
  content: '';
  left: 2px;
  bottom: 2px;
  background-color: white;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  border-radius: 50%;
  width: 16px;
  height: 16px;
}

.emotion-8:before {
  width: 16px;
  height: 16px;
}

.emotion-25 {
  font-size: 14px;
  color: #252626;
  padding-left: 8px;
  cursor: pointer;
  line-height: 18px;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 400;
}

<div>
    <label
      class="emotion-0"
    >
      <div
        class="emotion-1"
      >
        <input
          class="emotion-2"
          type="checkbox"
        />
        <span
          class="emotion-3 emotion-4"
        />
      </div>
    </label>
    <label
      class="emotion-0"
    >
      <div
        class="emotion-6"
      >
        <input
          class="emotion-7"
          type="checkbox"
        />
        <span
          class="emotion-8 emotion-4"
        />
      </div>
    </label>
    <label
      class="emotion-0"
    >
      <div
        class="emotion-1"
      >
        <input
          checked=""
          class="emotion-2"
          type="checkbox"
        />
        <span
          class="emotion-3 emotion-4"
        />
      </div>
    </label>
    <label
      class="emotion-0"
    >
      <div
        class="emotion-6"
      >
        <input
          checked=""
          class="emotion-7"
          type="checkbox"
        />
        <span
          class="emotion-8 emotion-4"
        />
      </div>
    </label>
    <label
      class="emotion-0"
    >
      <div
        class="emotion-6"
      >
        <input
          checked=""
          class="emotion-7"
          type="checkbox"
        />
        <span
          class="emotion-8 emotion-4"
        />
      </div>
      <div
        class="emotion-25"
      >
        Label text
      </div>
    </label>
  </div>
</DocumentFragment>
`;
